cpp/src/arrow/util/async_generator.h

```diff
@@ -1611,4 +1611,48 @@ AsyncGenerator<T> MakeCancellable(AsyncGenerator<T> source, StopToken stop_token
   return CancellableGenerator<T>{std::move(source), std::move(stop_token)};
 }
 
+template <typename T>
+class OrGenerator {
```

Comment:
RxJS calls this [defaultIfEmpty](https://www.learnrxjs.io/learn-rxjs/operators/conditional/defaultifempty) which is a bit more explicit.  My first thought when I see OrGenerator would be something like... `AsyncGenerator<X>(AsyncGenerator<optional<X>>, X default)`

Reply:
Ah, I was trying to come up with a good name. I'll change to something along those lines. (I was looking at Rust's Option::or which is a little too terse in this context.)